test_unstructured/partition/html/test_parser.py

```diff
@@ -0,0 +1,600 @@
+
+    The `Pre` class specializes behaviors for the `<pre>` (pre-formatted text) element.
+    """
+
+    def it_preserves_the_whitespace_of_its_phrasing_only_contents(self):
+        """A `<pre>` element can contain only phrasing content."""
+        html_text = (
+            "<pre>\n"
+            "  The Answer to the Great Question...   Of Life, the Universe and Everything...\n"
+            "  Is... Forty-two, said Deep Thought, with infinite majesty and calm.\n"
+            "</pre>\n"
+        )
+        pre = etree.fromstring(html_text, html_parser).xpath(".//pre")[0]
+
+        elements = pre.iter_elements()
+
+        e = next(elements)
+        assert e == Text(
+            "  The Answer to the Great Question...   Of Life, the Universe and Everything...\n"
+            "  Is... Forty-two, said Deep Thought, with infinite majesty and calm."
+        )
+        with pytest.raises(StopIteration):
+            next(elements)
+
+    @pytest.mark.parametrize(
+        ("html_text", "expected_value"),
+        [
+            # -- a newline in the 0th position of pre.text is dropped --
+            ("<pre>\n  foo  </pre>", "  foo  "),
+            # -- but not when preceded by any other whitespace --
+            ("<pre> \n  foo  </pre>", " \n  foo  "),
+            # -- and only one is dropped --
+            ("<pre>\n\n  foo  </pre>", "\n  foo  "),
+            # -- a newline in the -1th position is dropped --
+            ("<pre>  foo  \n</pre>", "  foo  "),
+            # -- but not when followed by any other whitespace --
+            ("<pre>  foo  \n </pre>", "  foo  \n "),
+            # -- and only one is dropped --
+            ("<pre>  foo  \n\n</pre>", "  foo  \n"),
+            # -- a newline in both positions are both dropped --
+            ("<pre>\n  foo  \n</pre>", "  foo  "),
+            # -- or not when not at the absolute edge --
+            ("<pre> \n  foo  \n </pre>", " \n  foo  \n "),
+        ],
+    )
+    def but_it_strips_a_single_leading_or_trailing_newline(
+        self, html_text: str, expected_value: str
+    ):
+        """Content starts on next line when opening `<pre>` tag is immediately followed by `\n`"""
+        pre = etree.fromstring(html_text, html_parser).xpath(".//pre")[0]
+        e = next(pre.iter_elements())
+
+        assert e.text == expected_value
+
+    def it_assigns_emphasis_and_link_metadata_when_contents_have_those_phrasing_elements(self):
+        html_text = '<pre>You\'re <b>turning</b> into a <a href="http://eie.io">penguin</a>.</pre>'
+        pre = etree.fromstring(html_text, html_parser).xpath(".//pre")[0]
+
+        e = next(pre.iter_elements())
+
+        assert e.text == "You're turning into a penguin."
+        assert e.metadata.emphasized_text_contents == ["turning"]
+        assert e.metadata.emphasized_text_tags == ["b"]
+        assert e.metadata.link_texts == ["penguin"]
+        assert e.metadata.link_urls == ["http://eie.io"]
+
+
```

Comment:
Should there be tests for TableBlock?

Reply:
`TableBlock` is essentially a transplant from `HTMLDocument`, so it still has all the tests that it did there. I purposely didn't make any changes to that implementation so I didn't see the need to add more tests.

We'll probably be back in there at some time and if so that would be a good time to translate some of those integration tests into isolated unit tests and maybe add some more.